src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala

```diff
@@ -115,12 +115,19 @@ class DeploymentActor(
     promise.future
   }
 
-  def scaleApp(app: AppDefinition, scaleTo: Int): Future[Unit] = {
+  def scaleApp(app: AppDefinition, scaleTo: Int, toKill: Option[Set[MarathonTask]]): Future[Unit] = {
     val runningTasks = taskTracker.get(app.id)
-    if (scaleTo == runningTasks.size) {
-      Future.successful(())
+    val runningTasksToKill = DeploymentDeathRow(runningTasks, toKill.getOrElse(Set.empty), scaleTo).tasksToBeKilled
```

Comment:
This could be a simple function call.

I'd actually like to move as much logic out of this function in the DeploymentActor as possible into a pure function outside.

It could just return the Set of tasks to kill (might be empty or you model that with an Option[Set]) and an Option[Int]  scaleToOpt.